### DISCUSSION TITLE
[21년 05월 25일] 피어세션 내용

### DISCUSSION THREAD

**Discussion by sglee487** [bcaitech1/p4-mod-model_diet#10] [General]
* 메탈 폴더 안에 그냥 보면 안보이는데 listdir로 보면 안에 .ipynb가 보인다. 아마 jupyter notebook으로 이미지 파일을 봐서 .ipynb 가 생긴 것 같다.

```
data_path = os.path.join('/','opt','ml','input','data')
for i in glob(os.path.join(data_path,'train','')+'*'):
    j=i+'/.ipynb_checkpoints'
    !rm -rf $j
```
폴더 내에 .ipynb_checkpoints 지우기 (편집됨) 

* Ustage에서 경량화 시간에 배웠던 model 구조까지 바꾸는건 Optuna에 없을까? 수업 후반부 쯤에 배우지 않을까..

 * FP16도 연산량 안쓰는게 정확도는 높이면서 연산량은 그대로니까
어차피 연산 횟수는 그대로이고 복잡도를 따지는게 아니니까. 수업에서는 int32 로 바꾸는걸로 했지만 float으로 해도 정확도를 올릴 수 있지 않을까.
FP16은 dataset config에서 False로 바꾸면 된다.

* timm 사용법.https://github.com/rwightman/pytorch-image-models
근데 pretrained 불러오려면 train.py를 싹다 뜯어 고쳐야 한다.

* cross entropy + Focal loss
```
# https://discuss.pytorch.org/t/is-this-a-correct-implementation-for-focal-loss-in-pytorch/43327/8
class FocalLoss(nn.Module):
    def __init__(self, weight=None,
                 gamma=2., reduction='mean', **kwargs):
        nn.Module.__init__(self)
        self.weight = weight
        self.gamma = gamma
        self.reduction = reduction
    def forward(self, input_tensor, target_tensor):
        log_prob = F.log_softmax(input_tensor, dim=-1)
        prob = torch.exp(log_prob)
        # https://wikidocs.net/60572
        return F.nll_loss(
            ((1 - prob) ** self.gamma) * log_prob,
            target_tensor,
            weight=self.weight,
            reduction=self.reduction
        )
class Cross_FocalLoss(nn.Module):
    def __init__(self, weight=None,
                 gamma=2., reduction='mean', **kwargs):
        nn.Module.__init__(self)
        self.weight = weight
        self.gamma = gamma
        self.reduction = reduction
    def forward(self, inputs, targets):
        # cross entropy
        loss_fct_cross = torch.nn.CrossEntropyLoss()
        loss_cross = loss_fct_cross(inputs, targets)
        # focal
        loss_fct_focal = FocalLoss()
        loss_focal = loss_fct_focal(inputs, targets)
        return loss_cross * 0.75 + loss_focal * 0.25
```
